src/MoveUntilTouchTopicController.cpp

```diff
@@ -0,0 +1,216 @@
+
+  // initialize base trajectory controller
+  return initController(robot, nh);
+}
+
+//=============================================================================
+void MoveUntilTouchTopicController::forceTorqueDataCallback(const geometry_msgs::WrenchStamped& msg)
+{
+  mForce.x() = msg.wrench.force.x;
+  mForce.y() = msg.wrench.force.y;
+  mForce.z() = msg.wrench.force.z;
+  mTorque.x() = msg.wrench.torque.z;
+  mTorque.y() = msg.wrench.torque.z;
+  mTorque.z() = msg.wrench.torque.z;
+}
+
+//=============================================================================
+void MoveUntilTouchTopicController::taringTransitionCallback(TareActionClient::GoalHandle goalHandle) {
+  if (goalHandle.getResult()) {
+    if (goalHandle.getResult()->success) {
+      ROS_INFO("Taring completed!");
+      mTaringCompleted.store(true);
+    }
+  }
+}
+
+//=============================================================================
+void MoveUntilTouchTopicController::starting(const ros::Time& time)
+{
+  // start asynchronous tare request
+  ROS_INFO("Starting Taring");
+  pr_control_msgs::TriggerGoal goal;
+  mTareGoalHandle = tareActionClient->sendGoal(goal, boost::bind(&MoveUntilTouchTopicController::taringTransitionCallback, this, _1));
+
+  // start base trajectory controller
+  startController(time);
+}
+
+//=============================================================================
+void MoveUntilTouchTopicController::stopping(const ros::Time& time)
+{
+  // stop base trajectory controller
+  stopController(time);
+}
+
+//=============================================================================
+void MoveUntilTouchTopicController::update(const ros::Time& time,
+                                      const ros::Duration& period)
+{
+  // update base trajectory controller
+  updateStep(time, period);
+}
+
+//=============================================================================
+bool MoveUntilTouchTopicController::shouldAcceptRequests() { return isRunning(); }
+
+//=============================================================================
+bool MoveUntilTouchTopicController::shouldStopExecution(std::string& reason)
+{
+  // inelegent to just terminate any running trajectory,
+  // but we must guarantee taring completes before starting
+  if (!mTaringCompleted.load()) {
+    ROS_WARN("taring not yet completed!");
+    return true;
+  }
+
+  double forceThreshold = mForceThreshold.load();
+  double torqueThreshold = mTorqueThreshold.load();
+
+  forceTorqueDataMutex.lock();
+  bool forceThresholdExceeded = mForce.norm() >= forceThreshold;
+  bool torqueThresholdExceeded = mTorque.norm() >= torqueThreshold;
+  forceTorqueDataMutex.unlock();
+
+  if (forceThresholdExceeded) {
+    reason = "Force Threshold exceeded!";
+    ROS_WARN(("Force Threshold exceeded!   " + std::to_string(forceThreshold) + "  |  " + std::to_string(mForce.x()) + " " + std::to_string(mForce.y()) + " " + std::to_string(mForce.z())).c_str());
```

Comment:
This more descriptive reason could be useful elsewhere too. Maybe we can use `stringstream` to fill out `reason` and then just `ROS_WARN(reason)`?